servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java

```diff
@@ -130,54 +140,73 @@ public void onSubscribe(final Cancellable cancellable) {
         }
 
         @Override
-        public void onSuccess(@Nullable final StreamingHttpResponse result) {
-            if (result == null || !shouldRedirect(redirectCount + 1, result, request.method())) {
+        public void onSuccess(@Nullable final StreamingHttpResponse response) {
+            if (response == null) {
+                target.onSuccess(null);
+                return;
+            }
 
-                target.onSuccess(result);
+            final boolean shouldRedirect;
+            try {
+                shouldRedirect = shouldRedirect(redirectCount + 1, request, response);
+            } catch (final Throwable cause) {
+                target.onError(cause);
+                return;
+            }
+            if (!shouldRedirect) {
+                target.onSuccess(response);
                 return;
             }
 
             final StreamingHttpRequest newRequest;
             try {
-                newRequest = prepareRedirectRequest(request, result, redirectSingle.requester);
+                newRequest = prepareRedirectRequest(request, response, redirectSingle.requester);
             } catch (final Throwable cause) {
                 target.onError(cause);
                 return;
             }
             if (newRequest == null) {
-                target.onSuccess(result);
+                target.onSuccess(response);
                 return;
             }
-            String newScheme = newRequest.scheme();
-
-            // Bail on the redirect if non-relative when that was requested or a redirect request is impossible to infer
-            if (redirectSingle.onlyRelative) {
-                HostAndPort oldEffectiveHostAndPort = request.effectiveHostAndPort();
-                HostAndPort newEffectiveHostAndPort = newRequest.effectiveHostAndPort();
-                if (oldEffectiveHostAndPort == null || !oldEffectiveHostAndPort.equals(newEffectiveHostAndPort)) {
-                    if (LOGGER.isDebugEnabled()) {
-                        LOGGER.debug("Ignoring non-relative redirect to '{}' for original request '{}': {}",
-                                newRequest.requestTarget(), redirectSingle.originalRequest,
-                                "Only relative redirects are allowed");
-                    }
-                    target.onSuccess(result);
-                    return;
-                } else if (newScheme != null) {
-                    // Rewrite absolute-form location to relative-form request-target in case only relative redirects
-                    // are supported and this is a relative redirect
+
+            final String newScheme = newRequest.scheme();
+            if (isRelative(request, scheme, newRequest)) {
+                if (redirectSingle.allowNonRelativeRedirects && newScheme == null && scheme != null) {
```

Comment:
The comment one line below describes motivation. Because this filter can be applied at a multi-address level when `allowNonRelativeRedirects == true`, we have to meet its requirements. Multi-address client requires users to provide full URL (even for relative redirects) because it always parses the request-target to identify which single-address client to use.

Reply:
Thanks for the explanation. Makes sense.